pkg/personname/info.go

```diff
@@ -0,0 +1,150 @@
+/*
+The personName package provides methods and data types for inspecting Person Name (PN)
+Value Representations, as defined here:
+
+http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_6.2.html
+*/
+package personname
+
+import (
+	"strings"
+)
+
+const groupSep = "="
+
+// Info holds information from an element with a "PN" VR. See the "PN"
+// entry at: http://dicom.nema.org/dicom/2013/output/chtml/part05/sect_6.2.html
+//
+// PN values are split into three groups which represent three different ways to
+// represent a name:
+//
+// - Alphabetic: How a name is formally spelled using a Phonetic alphabet.
+// - Ideographic: How a name is represented using ideograms / ideographs.
+// - Phonetic: How a name is pronounced.
+//
+// Each of these groups can be inspected to access their individual segments (family
+// name, Given name, etc.)
+type Info struct {
+	// Alphabetic group information about the Alphabetic group.
+	Alphabetic GroupInfo
+	// Ideographic group information about the Ideographic group.
+	Ideographic GroupInfo
+	// Phonetic group information about the Phonetic group.
+	Phonetic GroupInfo
+
+	// NoNullSeparators will remove repeated separators around null groups when
+	// calling String() if set to true.
+	NoNullSeparators bool
+}
+
+// WithNullSeparators returns a new Info object that will keep trailing separators
+// that surround both null groups AND group segments: (ex: 'Potter^Harry^^^==').
+func (info Info) WithNullSeparators() Info {
+	info.NoNullSeparators = false
+	info.Alphabetic.NoNullSeparators = false
+	info.Ideographic.NoNullSeparators = false
+	info.Phonetic.NoNullSeparators = false
+
+	return info
+}
+
+// WithoutNullSeparators returns a new Info object that will remove trailing
+// separators that surround both null groups AND group segments:
+// (ex: 'Potter^Harry').
+func (info Info) WithoutNullSeparators() Info {
+	info.NoNullSeparators = true
+	info.Alphabetic.NoNullSeparators = true
+	info.Ideographic.NoNullSeparators = true
+	info.Phonetic.NoNullSeparators = true
+
+	return info
+}
+
+// String returns the original Raw representation of the PN value, in
+// '[Alphabetic]=[Ideographic]=[Phonetic]' format.
+func (info Info) String() string {
+	groupStrings := []string{
+		info.Alphabetic.String(), info.Ideographic.String(), info.Phonetic.String(),
+	}
+
+	// If we are removing trailing emtpy, we are going to check if the last group is
+	// empty, then trim it off the end if it is, until we hit a non-empty slice.
+	if info.NoNullSeparators {
+		out := 3
+		for i := 2; i >= 0; i-- {
+			if groupStrings[i] == "" {
+				out = i
+			} else {
+				break
+			}
+		}
+
+		// Trim the groups by getting a slice of our slice.
+		groupStrings = groupStrings[0:out]
+	}
+
+	return strings.Join(groupStrings, groupSep)
```

Comment:
I think this can be simplified to be more readable (let me know if this is what you were going for):
```suggestion
	groupStrings := []string{
		info.Alphabetic.String(), info.Ideographic.String(), info.Phonetic.String(),
	}
	var sb strings.Builder
	for _, s := range groupStrings {
		if info.NoNullSeparators && s == "" {
			break
		}
		sb.WriteString(s)
		sb.WriteString(groupSep)
	}

	return sb.String()
```

Reply:
The issue here is that you need to inspect the last group first, and remove said group if it is empty.

If we check front to back, and middle second element is blank, but the last element has data, then bailing on the second element will give this:

"Harry^Potter"

Instead of this "Harry^Potter==Phonetic^Spelling", which is a valid PN value. Point taken on the code being a little ugly. I'll see if I can clean it up.

Reply:
To add to this: We could invert the slice ahead of time, but then we have to loop over it and build a new slice even if we are not culling null separators.

Reply:
Ive opted to move the loop value-culling logic into a clearly named helper function, and add more detailed code comments about why the backwards iteration approach is being taken.

Leaving this open for your thoughts after taking a look.

Reply:
Side note: I did not know about the strings.Builder type, so thanks for teaching me something new todday!

Reply:
Ah I see your use case. I think this is reasonable for now, thank you for the refactor!